beluga/include/beluga/sensor/ndt_sensor_model.hpp

```diff
@@ -0,0 +1,234 @@
+#include <stdexcept>
+#include <type_traits>
+#include <unordered_map>
+#include "H5Cpp.h"
+
+namespace beluga {
+
+// Hash function for Eigen::Vector2i.
+// The code is from `hash_combine` function of the Boost library. See
+// http://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine .
+struct CellHasher {
+  /// Hashes an integer 2D vector.
+  size_t operator()(const Eigen::Vector2i& vector) const {
+    size_t seed = 0;
+    for (auto i = 0L; i < vector.size(); ++i) {
+      auto elem = *(vector.data() + i);
+      seed ^= std::hash<int>()(elem) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+    }
+    return seed;
+  }
+};
+
+/// Parameters used to construct a NDTSebsorModel instance.
+struct NDTModelParam {
+  /// Likelihood for measurements that lie inside cells that are not present in the map.
+  double minimum_likelihood = 0;
+  /// Scaling parameter d1 in literature, used for scaling 2D likelihoods.
+  double d1 = 1.0;
+  /// Scaling parameter d2 in literature, used for scaling 2D likelihoods.
+  double d2 = 1.0;
+};
+
+/// Representation for a cell of a 2D NDT map.
+struct NDTCell {
+  /// Mean of the 2D normal distribution.
+  Eigen::Vector2d mean;
+  /// Covariance of the 2D normal distribution.
+  Eigen::Matrix<double, 2, 2> covariance;
+
+  /// Get the L2 likelihood at measurement, scaled by d1 and d2. It assumes the measurement is pre-transformed
+  /// into the same frame as this cell instance.
+  [[nodiscard]] double likelihood_at(const NDTCell& measurement, double d1 = 1.0, double d2 = 1.0) const {
+    const Eigen::Vector2d error = measurement.mean - mean;
+    const double rhs =
+        std::exp((-d2 / 2.0) * error.transpose() * (measurement.covariance + covariance).inverse() * error);
+    return d1 * rhs;
+  }
+};
+
+/// Transform the NDT according to tf, both mean and covariance.
+inline NDTCell operator*(const Sophus::SE2d& tf, const NDTCell& ndt_cell) {
+  const Eigen::Vector2d uij = tf * ndt_cell.mean;
+  const Eigen::Matrix2Xd cov = tf.so2().matrix() * ndt_cell.covariance * tf.so2().matrix().transpose();
+  return NDTCell{uij, cov};
+}
+
+/// Fit a vector of points to an NDT cell, by computing its mean and covariance.
+inline NDTCell fit_points(const std::vector<Eigen::Vector2d>& points) {
+  static constexpr double kMinVariance = 1e-5;
+  Eigen::Map<const Eigen::Matrix2Xd> points_view(
+      reinterpret_cast<const double*>(points.data()), 2, static_cast<int64_t>(points.size()));
+  Eigen::Vector2d mean = points_view.rowwise().mean();
+  Eigen::Matrix2Xd centered = points_view.colwise() - mean;
+  // Use sample covariance.
+  Eigen::Matrix2Xd cov = (centered * centered.transpose()) / static_cast<double>(points_view.cols() - 1);
+  cov(0, 0) = std::max(cov(0, 0), kMinVariance);
+  cov(1, 1) = std::max(cov(1, 1), kMinVariance);
+  return NDTCell{mean, cov};
+}
+
+/// Ostream overload mostly for debugging purposes.
+inline std::ostream& operator<<(std::ostream& os, const NDTCell& cell) {
```

Comment:
@serraramiro1 consider relocating operator overloads to class / struct scope as inline friend functions, so as to limit them to ADL resolution (which means less verbosity from the compiler when an error occurs).

Reply:
Nice one. Done.